calculation/spatial_hazard/spatial_hazard/correlate_ims.py

```diff
@@ -0,0 +1,227 @@
+        emp_df["mu"],
+        emp_df["sigma_inter"],
+        emp_df["sigma_intra"],
+    )
+
+    # Decompose the covariance matrix
+    # L = np.linalg.cholesky(nearest_pd(R))
+
+    # Cholesky decomposition of the correlation matrix
+    # Make positive definite if it isn't
+    try:
+        L = cholesky(R, lower=True)
+    except np.linalg.LinAlgError:
+        pd_R = nearest_pd(R)
+        L = cholesky(pd_R, lower=True)
+
+    random_IMs = np.zeros((N, len(stations)))
+    between_event = np.zeros((N, len(stations)))
+    within_event = np.zeros((N, len(stations)))
+
+    for i in range(N):
+        # Compute one single between event for everyone
+        normalized_between_event = np.random.normal(0.0, 1.0)
+        between_event_term = normalized_between_event * between_event_std
+
+        # correlate the within event term using L (Box-Muller)
+        within_event_term = np.matmul(L, np.random.normal(0.0, within_event_std))
+
+        # We now have a perturbated mean field
+        IM = mean_lnIM + between_event_term + within_event_term
+
+        # Add it to the list
+        random_IMs[i, :] = IM
+        between_event[i, :] = between_event_term
+        within_event[i, :] = within_event_term
+
+    return random_IMs, between_event, within_event
+
+
+def nearest_pd(A):
+    """Find the nearest positive-definite matrix to input
+
+    A Python/Numpy port of John D'Errico's `nearestSPD` MATLAB code [1], which
+    credits [2].
+
+    [1] https://www.mathworks.com/matlabcentral/fileexchange/42885-nearestspd
+
+    [2] N.J. Higham, "Computing a nearest symmetric positive semidefinite
+    matrix" (1988): https://doi.org/10.1016/0024-3795(88)90223-6
+    """
+    B = (A + A.T) / 2
+    _, s, V = np.linalg.svd(B)
+
+    H = np.dot(V.T, np.dot(np.diag(s), V))
+
+    A2 = (B + H) / 2
+
+    A3 = (A2 + A2.T) / 2
+
+    if is_pd(A3):
+        return A3
+
+    spacing = np.spacing(np.linalg.norm(A))
+    # The above is different from [1]. It appears that MATLAB's `chol` Cholesky
+    # decomposition will accept matrixes with exactly 0-eigenvalue, whereas
+    # Numpy's will not. So where [1] uses `eps(mineig)` (where `eps` is Matlab
+    # for `np.spacing`), we use the above definition. CAVEAT: our `spacing`
+    # will be much larger than [1]'s `eps(mineig)`, since `mineig` is usually on
+    # the order of 1e-16, and `eps(1e-16)` is on the order of 1e-34, whereas
+    # `spacing` will, for Gaussian random matrixes of small dimension, be on
+    # othe order of 1e-16. In practice, both ways converge, as the unit test
+    # below suggests.
+    I = np.eye(A.shape[0])
+    k = 1
+    while not is_pd(A3):
+        mineig = np.min(np.real(np.linalg.eigvals(A3)))
+        A3 += I * (-mineig * k ** 2 + spacing)
+        k += 1
+
+    return A3
+
+
+def is_pd(B):
+    """Returns true when input is positive-definite, via Cholesky"""
+    try:
+        _ = cholesky(B, lower=True)
+        return True
```

Comment:
Does this mean if `cholesky()` function works with no issue, we return True otherwise False?

Just curious about those two lines haha and also whether these could be simplified.